source/fundamentals/csfle.txt

```diff
@@ -0,0 +1,289 @@
+==================================
+Client-Side Field Level Encryption
+==================================
+
+.. default-domain:: mongodb
+
+Client-Side Field Level Encryption (CSFLE) allows you to encrypt
+specific data fields within a document with your MongoDB client application before sending the data to the server.
+Starting in MongoDB 4.2 Enterprise, you can perform this client-side encryption automatically.
+
+With CSFLE, your client application encrypts fields client-side without requiring any server-side configuration or
+directives. CSFLE is useful for situations in which applications must guarantee that unauthorized parties, including
+server administrators, cannot read the encrypted data.
+
+This guide is a quick introduction to CSFLE using the Java driver. For in-depth information on how CSFLE works, see
+the :manual:`CSFLE reference </core/security-client-side-encryption/>` documentation. For a real-world scenario and
+implementation, see our `CSFLE Guide <https://docs.mongodb.com/drivers/security/client-side-field-level-encryption-guide>`_.
+
+Installation
+------------
+
+To get started with CSFLE in your client application, you need
+
+- the MongoDB Java driver
+- ``libmongocrypt``
+- ``mongocryptd`` if using automatic encryption (Enterprise & Atlas)
+
+``libmongocrypt``
+~~~~~~~~~~~~~~~~~
+
+The ``libmongocrypt`` binding is available as a separate JAR. Add it to your project using your desired dependency
+management tool.
+
+.. tabs::
+
+   .. tab:: Maven
+      :tabid: maven
+
+      .. include:: /includes/fundamentals/code-snippets/libmongocrypt-maven-versioned.rst
+
+   .. tab:: Gradle
+      :tabid: gradle
+
+      .. include:: /includes/fundamentals/code-snippets/libmongocrypt-gradle-versioned.rst
+
+
+``mongocryptd``
+~~~~~~~~~~~~~~~
+
+``mongocryptd`` is a binary run as a daemon / process that is used for automatic encryption.
+``libmongocrypt`` communicates with ``mongocryptd`` to automatically encrypt the information specified in a user-provided
+:manual:`JSON Schema </reference/security-client-side-automatic-json-schema/>`.
+
+For more detailed information on ``mongocryptd``, see
+:manual:`mongocryptd reference documentation </reference/security-client-side-encryption-appendix/#mongocryptd>`
+
+
+Examples
+--------
+
+The examples on this page use a local key, but you can also use integrated support for cloud-based key management
+services from AWS, Azure, and GCP. Each example program execution creates a new master key in memory and drops the
+``test.coll`` collection.
```

Comment:
The local master key examples make it difficult to summarize the behavior since it performs the following:
-creates a new master key
-creates a new data encryption key using that master key (meaning it's only valid for that run)
-drops the test.coll collection in case there were any other docs in there
-inserts a new document
-retrieves and prints the document
-leaves as a side-effect: the encrypted document and the data key, but not the master key. Therefore, you can only view the encrypted document as you need the master key to decrypt the data encryption key to decrypt the document fields.

Therefore, we should warn the user that they cannot decrypt the fields once the master key reference is lost. Maybe we can accomplish this by appending something like the following:

"
.. warning:
   In the examples, the in-memory master key is lost when the application finishes running. If you want to decrypt the document fields in a separate run, you can save the local master key to a file and reuse it. For use In a production environment, we recommend that you save the key to a key management service.
"